include/phasar/Utils/SemiRing.h

```diff
@@ -0,0 +1,34 @@
+/******************************************************************************
+ * Copyright (c) 2024 Fabian Schiebel.
+ * All rights reserved. This program and the accompanying materials are made
+ * available under the terms of LICENSE.txt.
+ *
+ * Contributors:
+ *     Fabian Schiebel and others
+ *****************************************************************************/
+
+#ifndef PHASAR_UTILS_SEMIRING_H
+#define PHASAR_UTILS_SEMIRING_H
+
+#include "phasar/DataFlow/IfdsIde/EdgeFunction.h"
+
+namespace psr {
+template <typename AnalysisDomainTy> class SemiRing {
```

Comment:
Hmm, as the IDESolver is already specialized on the `AnalysisDomainTy`, could it make sense to implement the `SemiRing` structure as a trait and not on the ide/ifds problem itself.

Question: Is the implementation of combine/extend usually/always problem specific or is it in general specific for a given `AnalysisDomainTy`.

Why do I ask? Well, one would model the interface around a trait, the IDESolver would get around the virtual function call to extend/combine.

Example:
```cpp
...
template<>
class SemiRing<MyAnalysisDomain> {
  static EdgeFunction<l_t> extend(const EdgeFunction<l_t> &L,
                                                       const EdgeFunction<l_t> &R) {
    ...
  }
}

class IDESolver {
  void foo() {
    ...
    auto res = SemiRing<AnalysisDomainTy>::extend(L, R);
    ...
  }
};
```

Reply:
I haven't benchmarked the new implementation, but I like your idea with separating the semi ring from the problem.
However, we cannot implement the trait functions as `static` as the main reason for this PR is to make the integration of edge-function caching easier. Still, we can get rid of the additional virtual call

Reply:
There is one problem, though: The edge-function factories within the IDE problem need the edge-function cache, as well as the `extend` and `combine` functions need it.
So, it is actually easier to have these both functions as members within the IDE problem; or do you have an idea, how to workaround this?

Reply:
Why can the function not be static for the cache? The cache state could be part of the IDE solver and in case of a miss call the static impl. but in the end, solve it as you like it best^^

Reply:
I think, the edge functions should be specific to the problem. Two problems might accidentially have the same domain, but want to handle edge functions differently.

The idea of having the cache in the solver sounds interesting, but probably too much for this PR.